### ISSUE TITLE
Upcoming release requirements

### ISSUE THREAD

**Issue by aparcar** [openwrt/openwrt#9248]
This is a meta issue to keep track of what's needed for the next release.

- [ ] All targets must be on Kernel 5.10
  - #5048 
- [ ]  `firewall4` must work
  - https://github.com/openwrt/packages/issues/16818
  - https://github.com/openwrt/luci/issues/5409
  - https://github.com/openwrt/routing/issues/731
- [ ] more

All project members may edit this issue and add missing features/bugs

**Comment by nicklowe** [comment]
With WPA3, I suggest enabling H2E for SAE before forking: https://github.com/openwrt/openwrt/pull/9300

**Comment by Qosifyck** [comment]
Please @aparcar, can you migrate OpenWrt to [odhcpd + unbound](https://openwrt.org/docs/guide-user/services/dns/dot_unbound#command-line_instructions) by default and remove dnsmasq (because it's garbage) and if you do it would be the biggest change (for better) in OpenWrt since its founding.

@aparcar you once asked it in the forum:
- [Odhcpd by default? - OpenWrt Forum](https://forum.openwrt.org/t/odhcpd-by-default/92704)

**Comment by jow-** [comment]
We will not replace dnsmasq with odhcpd + unbound at this point in time.

**Comment by Qosify** [comment]
@aparcar I don't understand why you don't want to replace dnsmasq with [odhcpd + unbound](https://openwrt.org/docs/guide-user/base-system/dhcp_configuration#replacing_dnsmasq_with_odhcpd_and_unbound), if Unbound is much better and also encrypts DNS by default without having to install external packages and offers better performance than dnsmasq and all the web pages will load super fast 100% verified.

**Comment by PolynomialDivision** [comment]
Would have been nice to have https://github.com/openwrt/openwrt/pull/4721 included. (cc @robimarko) However, with a feature freeze on 1 March, we would need to be very quick converting the last boards. ;)

**Comment by PolynomialDivision** [comment]
Sysupgrade is broken on ath79 ubnt-xm, due to oom-killer killing sysupgrade.

We have two options:
1) Finding more processes to kill before doing sysupgrade
2) Move ath79 ubnt-xm to tiny like I did in the PR: https://github.com/openwrt/openwrt/pull/4879

**Comment by champtar** [comment]
@PolynomialDivision any manual workaround ? I still have a bunch of them in the field :(

**Comment by robimarko** [comment]
> Would have been nice to have #4721 included. (cc @robimarko) However, with a feature freeze on 1 March, we would need to be very quick converting the last boards. ;)

Unfortunately, I don't think this is viable before the feature freeze and it should not be rushed.

**Comment by reusu** [comment]
fullcone nat support？

**Comment by varoudis** [comment]
#4996 maybe? arm64 with UEFI.

**Comment by jow-** [comment]
> fullcone nat support？
Just add static SNAT and DNAT rules?

**Comment by nicklowe** [comment]
We ought to also cover enabling H2E SAE in the case that wpa_supplicant is used as a client. I will create a patch.

**Comment by DarknessOpen** [comment]
> Just add static SNAT and DNAT rules?

Is not the same.

I would also like Full cone NAT support.

In stock firmware TOTOLINK A8000RU has Full cone NAT support:
![10f4c967d883575216d95da89aa9966489b1f387](https://user-images.githubusercontent.com/100588977/156191444-36ae8eb0-8821-474f-bacc-cf92481f2876.png)

* [Gaming with Full Cone vs Symmetric NAT Routers](https://web.archive.org/web/20210506124657/https://badmodems.com/Forum/viewtopic.php?t=21)
> For some voice and video concurrent services or games, there might be some connection issue with symmetric NAT. You might switch to full-cone NAT for a better connection. However, with full-cone all external host can send packets to intranet.

> "A full cone NAT is one where all requests from the same internal IP address and port are mapped to the same external IP address and port. Furthermore, any external host can send a packet to the internal host, by sending a packet to the mapped external address."

> Single NAT is where either a modem/router device or a modem only and a external router, where either the modem/router or external router handles all routing and NAT traffic and is the main host router for all connected devices including game consoles and PCs. The ISP public IP address is translated from the ISP modem to the gaming device seamlessly. WAN to LAN. WAN side ISP public IP address ##.##.###.### to a LAN side IP address of 192.168.#.#. The WAN side ISP public IP address ##.##.###.### is correctly translated to game consoles/PCs by just the external router handling the NAT traffic.
This configuration is ideal and what works best for gaming configurations and should be configured as such and is the "golden" configuration for general home networking and gaming environments.

> MS and Sony may have not fully tested two or more game console/PC and same or different game environments. Current marketing perspective maybe thinking of one household, one game console. uPnP code handling between the consoles and games are problematic. Router Mfr use of Symmetric NAT type only inhibits good NAT handling for two or more game consoles/PCs, where FULL CONE NAT handles two or more game consoles/PCs playing the same game at the same time correctly. Also games may have specific problems in handling of NAT when being played on routers with Symmetric NAT as well which should be looked at for multiple game environments by game developers. There also could be a problem with Symmetric NAT in itself as well which should be looked at. Can Symmetric NAT handle two or more game consoles? Again, all these considerations are only seen when using two or more game consoles. A mixed of consoles would also be a consideration as well, i.e. Xbox and Play station on the same router online at the same time.
>
>Game Mfrs also have a role to play in this and from what i've seen, some games are not handling uPnP ports and NAT kinds correctly with two or more consoles running same game, which needs to be addressed by game Mfrs which use and give NAT status in there game. I'm seeing that recently that two consoles running the same game are achieving OPEN NAT on Port Address Restricted NAT routers on the consoles network settings with two consoles, however after loading the game, one game console reports OPEN NAT and the other console reports MODERATE NAT. This is with the new MS xbox one port address change feature that was released fully on Oct 17th when it hit my xbox one console. When using a FULL CONE NAT router with two same game running consoles, OPEN NAT is seen across the board and is not an issue.

**Comment by champtar** [comment]
Many protocols figured out NAT traversal and full cone NAT is not present in OpenWrt since forever so this can't be a release blocker, this is a normal feature request.

**Comment by jow-** [comment]
"A full cone NAT is one where all requests from the same internal IP address and port are mapped to the same external IP address and port. Furthermore, any external host can send a packet to the internal host, by sending a packet to the mapped external address."

Sounds like a static SNAT/DNAT setup to me.

**Comment by DarknessOpen** [comment]
> Sounds like a static SNAT/DNAT setup to me.

I am sorry to tell you that it is not and due to your wrong thinking you have not implemented full cone NAT support in OpenWrt.

People have been asking/pleading for years to add Full cone NAT support by default in OpenWrt, but you guys have been ignoring them.

**Comment by DarknessOpen** [comment]
> Sounds like a static SNAT/DNAT setup to me.

Sorry to tell you that you are wrong and due to your wrong thinking you have not implemented Full cone NAT support in OpenWrt.

People have been asking/pleading for years to add Full cone NAT support by default in OpenWrt, but you guys have been ignoring them.

**Comment by jow-** [comment]
> I am sorry to tell you this, but you are wrong and due to your wrong thinking you have not implemented Full cone NAT support in OpenWrt.

What is wrong about the statement that a static SANT and DNAT rule achieve "full cone NAT" properties? What exactly makes "full cone NAT" different from a static SNAT and DNAT rule? Autoconfiguration?

**Comment by DarknessOpen** [comment]
> What exactly makes "full cone NAT" different from a static SNAT and DNAT rule? Autoconfiguration?

I'm going to go to your house and show you the difference with my fists.

**Comment by DarknessOpen** [comment]
> What exactly makes "full cone NAT" different from a static SNAT and DNAT rule? Autoconfiguration?

I'm going to go to your house and show you the difference with my fists.
![tumblr_m0upkib0GL1rqfhi2o1_500](https://user-images.githubusercontent.com/100588977/156361209-8a300aa5-f6bf-46e8-a710-3c8eda5072aa.gif)

**Comment by reusu** [comment]
> > I am sorry to tell you this, but you are wrong and due to your wrong thinking you have not implemented Full cone NAT support in OpenWrt.
> 
> What is wrong about the statement that a static SANT and DNAT rule achieve "full cone NAT" properties? What exactly makes "full cone NAT" different from a static SNAT and DNAT rule? Autoconfiguration?

i think 
masquerade = symmetric nat
snat + dnat = port forward != fullcone nat
that's all

**Comment by jow-** [comment]
> snat + dnat = port forward != fullcone nat

That still does not explain what "fullcone NAT" *is*. From what I could gather in the net, it is basically a form of NAT where a specific internal port of a specific internal host is mapped to a unique external port. A traditional port forward ensures that a given external port is forwarded to a specific internal host and port. Adding an additional SNAT rule to explicitly rewrite *outbound* traffic from the internal host and port to the specific external port without relying on MASQUERADE egress port auto allocation will achieve exactly that. What else is needed to achieve "full cone NAT" ?

**Comment by aparcar** [comment]
@jow- using my detective skills I found the following [code fragment](https://github.com/x-wrt/com.x-wrt/blob/09c93fd1ced8d477b9b0086593b3295799306382/natcap/files/natcapd.sh#L275):

```shell
		ipset create cone_wan_ip iphash hashsize 32 maxelem 256 >/dev/null 2>&1
		ipset flush cone_wan_ip
		devs=`ip r | grep default | grep -o "dev ".* | awk '{print $2}'`
		for dev in $devs; do
			ips=`ip addr list dev $dev | grep -o inet" ".* | awk '{print $2}' | cut -d/ -f1`
			for ip in $ips; do
				ipset add cone_wan_ip $ip >/dev/null 2>&1
			done
		done
```

**Comment by jow-** [comment]
@aparcar - and how is the `cone_wan_ip` set used?

**Comment by WinLinux1028** [comment]
There is a patch to use fullcone nat:
https://github.com/llccd/openwrt-fullconenat/tree/dev
However, it doesn't seem to be compatible with nftables.

**Comment by LA-CUCUCCUCURACHA** [comment]
In Asus routers you can also change the nat type:
![asus-rt-ac86u-upnp](https://user-images.githubusercontent.com/100805183/156453012-5ca3dfcd-a24b-4cfd-b115-9f655f149eae.jpg)

**Comment by LA-CUCUCCUCURACHA** [comment]
> There is a patch to use fullcone nat:
> https://github.com/llccd/openwrt-fullconenat/tree/dev
> https://github.com/LGA1150/openwrt-fullconenat
 > @llccd @LGA1150

In Asus routers you can also change the nat type:
![asus-rt-ac86u-upnp](https://user-images.githubusercontent.com/100805183/156453012-5ca3dfcd-a24b-4cfd-b115-9f655f149eae.jpg)

**Comment by LA-CUCUCCUCURACHA** [comment]
> There is a patch to use fullcone nat:
> https://github.com/llccd/openwrt-fullconenat/tree/dev
> https://github.com/LGA1150/openwrt-fullconenat

In Asus routers you can also change the nat type:
![asus-rt-ac86u-upnp](https://user-images.githubusercontent.com/100805183/156453012-5ca3dfcd-a24b-4cfd-b115-9f655f149eae.jpg)

@llccd @LGA1150 @jow- @aparcar

**Comment by LA-CUCUCCUCURACHA** [comment]
#7021

The default OpenWrt netfilter implements "Port Restricted Cone" NAT. Sometimes it's useful to have less restricted NAT.

Someone recently created a netfilter extension that implements "Full Cone" NAT. Please include it in the package feeds.

Full Cone NAT description:
https://tools.ietf.org/html/rfc3489#section-5

Netfilter extension:
https://github.com/Chion82/netfilter-full-cone-nat

OpenWrt package:

https://github.com/LGA1150/openwrt-fullconenat
https://github.com/LGA1150/fullconenat-fw3-patch

**Comment by mkresin** [comment]
Instead of posting script, links or insults one should go back and provide an answer to jows question

> What exactly makes "full cone NAT" different from a static SNAT and DNAT rule?"

With SNAT you do not know in advance which src port is used by the gateway to send the request (only high ports are used as src port and so on). Hence, you can not add a (static) DNAT rule for incoming traffic matching the gateways used src port.

Speaking in terms of SNAT and DNAT, with full cone NAT a DNAT rule is automatically added for the src port used by the gateway. DNAT dest port is the internal clients src port. I wouldn't be surprised if the pasted script mimics the behaviour.

I'm not entirely sure if with full cone NAT the gateway needs to use/uses the same src port as the internal client.

https://en.wikipedia.org/wiki/Network_address_translation#Methods_of_translation has an okayish illustration about the differences.

**Comment by champtar** [comment]
Another solution to the full cone NAT problem is switching to IPv6 ;)

**Comment by LA-CUCUCCUCURACHA** [comment]
IPv6 🤢🤮

**Comment by LA-CUCUCCUCURACHA** [comment]
> Instead of posting script, links or insults one should go back and provide an answer to jows question

The fools do not want to understand with words because they are deceived with their own lies due to their lack of knowledge and arrogance not recognizing that they are wrong and the only answer I can give to those lazy ones who do not want to learn are my fists.

![punch-beating](https://user-images.githubusercontent.com/100805183/156471452-06dba54e-5ea1-4d74-a2df-5bc3dec5876b.gif)

**Comment by predators46** [comment]
I can't wait for the 5.10.x kernel to be on the stable version of openwrt. i need musl version in snapshoot

**Comment by SorryJow** [comment]
I'm sorry @jow- for insulting you and threatening you.

![bart](https://user-images.githubusercontent.com/100843272/156554067-9b0ad226-be46-4620-b168-5fc64070f8df.gif)

**Comment by SorryJow** [comment]
Es tut mir leid @jow- 

https://user-images.githubusercontent.com/100845128/156558713-8ff9b62c-3188-4cb0-8b73-2e9867085248.mp4

**Comment by dfateyev** [comment]
Could we reduce the spam amount, and discuss here real blockers and requirements for the upcoming release.
This thread needs some moderation. Also, part of the messages can be wiped out entirely — they have no useful info and just spam.

**Comment by aparcar** [comment]
From my understanding we'll need the new dnsmasq version since it support nftables, could people please test the PR I linked?

**Comment by BabochkaAce** [comment]
[odhcpd](https://github.com/libremesh/lime-packages/issues/294) support nftables?

**Comment by Headcrabed** [comment]
> Another solution to the full cone NAT problem is switching to IPv6 ;)

Sadly, not every software supports ipv6, especially games. Besides, bug that ipv6 package loss when using flow offload haven
t been fixed yet.

**Comment by Borromini** [comment]
@aparcar I think https://github.com/openwrt/openwrt/issues/9377 'deserves' a place on this list as well, since it's a regression compared to 21.02.

**Comment by dfateyev** [comment]
@nicklowe you have mentioned H2E SAE support in AP client mode, do you have any plans to provide it?

**Comment by stangri** [comment]
I've asked this before and I'm sorry if I asked in the wrong issues/places.

My `vpn-policy-routing` and `vpnbypass` packages heavily depend on the `dnsmasq-full`'s ipset support. What's the plan for `dnsmasq-full` going forward, would it still include only `ipset` support in the next release or only `nftset` support or would it be split into `dnsmasq-full-ipset` and `dnsmasq-full-nftset`?

**Comment by nicefile** [comment]
@aparcar I would add regression [#3832](https://github.com/openwrt/openwrt/issues/8833) regarding switch ports with 10/100 Mbit PHY found in fritzbox 7360 7360v2 7362sl .lantiq xrx200 target

**Comment by BabochkaAce** [comment]
[odhcpd](https://github.com/libremesh/lime-packages/issues/294) or unbound support nftables?

**Comment by aparcar** [comment]
@BabochkaAce Please stop cross posting and repeating the same question again and again. It's unrelated to the release requirements. Please use the forum and open a new topic or issue here on GitHub.

**Comment by BabochkaAce** [comment]
Someone answered my question elsewhere.

As my Mother Russian says, "Я люблю спать, есть и какать" which means that Ukraine and all of Europe belong to Russia.

https://www.youtube.com/watch?v=_lJPr-L6gNo

**Comment by Ansuel** [comment]
@aparcar should we add this to the known issue? (it's like that from ages... considering wps push button is provided in luci, i think it should be noted)

https://github.com/openwrt/openwrt/issues/9369

**Comment by allhailCdosdude** [comment]
Is it possible, please to add regression[ 3832]( https://github.com/openwrt/openwrt/issues/8833)

**Comment by obsy** [comment]
Bug in qmi script: https://github.com/openwrt/openwrt/issues/9501

**Comment by robertfoss** [comment]
Ubiquiti Nanostations can no longer be flashed due to ram size limitations, a PR improving this situation has been submitted #4879.

**Comment by mpratt14** [comment]
I opened #9507 

probably just something to add to "known bugs"

I don't know if it can be fixed, but if so, it can wait until after branching anyway

**Comment by MaxSKash** [comment]
Enable `management` in openvpn

**Comment by mans0n** [comment]
I've just got to know that bcm53xx had a breaking change (convert to DSA) but did not bumped `DEVICE_COMPAT_VERSION`.
https://github.com/openwrt/openwrt/commit/e9672b1a8fa4714cbc35d5964caf704b2e571f35#commitcomment-58577826
The commit did add a simple migration script but I don't think that is enough... (especially for the stable release)
Ref: 47b531de3250 b1df48caac1b

**Comment by kofec** [comment]
One small issue: https://github.com/openwrt/openwrt/issues/9580

**Comment by howl** [comment]
#9676 should be considerer as the issue makes the connection to work very bad and is because a feature debuting in this release.

**Comment by Regloom** [comment]
Please cosider I2Pd update to 2.41 (mipselsf-k3.4 arch). Here is a link to issue [783](https://github.com/Entware/Entware/issues/783) on Entware.

**Comment by WinLinux1028** [comment]
#9881 I found a problem of 22.03(latest snapshot)

**Comment by Tabiskabis** [comment]
I used uci interfaces to map devices to firewall zones, and have OpenVPN configure those devices. This worked fine with fw3. 
fw4 however does not assign these devices to their zones. I worked around the issue by switching the interfaces protocol from "none" to "static" with no additional settings.
This behaviour seems faulty to me. In case that this is not considered a bug, it should at least get well documented.

**Comment by howl** [comment]
Another fw4 nftables issue #9923.

**Comment by spolack** [comment]
WLAN AP is broken on mt7622/mt7915 #9945

**Comment by howl** [comment]
I asked for consider a nftables bug but is not nftables, mt76 instead, but the impact is very severe openwrt/mt76#676.

**Comment by howl** [comment]
That issue seems to be caused by this commit in mt76 openwrt/mt76@e0f9479.

**Comment by spolack** [comment]
I would like to get the shortcoming of firewall4 #9981 handled, since it would break many use-cases.

**Comment by skbeh** [comment]
There is another patch to achieve fullcone nat:
https://github.com/fullcone-nat-nftables

**Comment by Shine-** [comment]
As far as I could see, there are still a few flow offload backports from >5.10 missing. Specifically [4592ee7f](https://github.com/torvalds/linux/commit/4592ee7f525c4683ec9e290381601fdee50ae110), which would at least fix #8776.

**Comment by Ansuel** [comment]
@Shine- seems easy enough to backport, Can you confirm that it does fix it?

**Comment by Shine-** [comment]
Well, I can at least confirm that my kernel 5.4 hack from #8776 works in 21.02. For 22.03, I haven't ever used a build _without_ the mentioned 5.15 backport. So while I can confirm "it works", I can't confirm that "it doesn't work" without the backport...

**Comment by msvamp** [comment]
Can we look into #5723 please? Seeing too many dnsmasq reloads occurring when multiple IPv6 devices are present, slowing down things without reason.

**Comment by attila-lendvai** [comment]
maybe this, too?

> miniupnpd litters the log with numerous different errors
https://github.com/openwrt/openwrt/issues/10267

**Comment by lespinasse** [comment]
Would https://github.com/openwrt/openwrt/issues/10243 qualify ? (MediaTek MT7610E 5GHz radio not functional in 22.03.0-rc5)

**Comment by rikka0w0** [comment]
The new BMIPS target adds support for the external MDIO bus, but if the DTS specifies both an internal and an external switch, the compiled image can no longer boot. Looks like the MDIO driver still has some problems.

https://github.com/openwrt/openwrt/issues/10313

**Comment by Borromini** [comment]
@rikka0w0 bmips is source only on 22.03 so it won't be part of this release.

**Comment by lespinasse** [comment]
#10243 still broken in the rc6 release (MediaTek MT7610E 5GHz radio not functional)

**Comment by andriej** [comment]
#8183 still broken - sysupgrade bricks NGB6716 while fix is there waiting

**Comment by lister-wrt** [comment]
[#10374](https://github.com/openwrt/openwrt/issues/10374) this ath79 device is broken and should probably be excluded from 22.03

**Comment by Lanchon** [comment]
@jow- 

> That still does not explain what "fullcone NAT" _is_.

sorry to gravedig. the 'coneness' of NATs is pretty arbitrarily defined. the concept lacks focus and utility, and it has been unclear to me until recently. this article does a good job of explaining, please skip to the middle and read the short "Naming our NATs" section:

https://tailscale.com/blog/how-nat-traversal-works/

a few takeaways:

- EDM NATs (symmetric NATs or hard NATs) are hard to hole-punch.
- EIM NATs (the rest) are easy, regardless of coneness. (but all require hole-punching nonetheless.)
- coneness is an abandoned taxonomy that has been replaced by newer more useful ones.

my current thinking:

- some feature implemented (or property displayed) by some router is not reason enough to implement it in openwrt: it has to be useful and its value must outweigh it maintenance cost.
- configurable coneness is not really useful:
  - it is an abandoned concept.
  - hole-punching done correctly is oblivious to coneness.
- while full cone might make some incapable apps traverse NATs in certain specific situations, it would be a hack to accommodate buggy apps.
- full cone does not solve any problem that cannot or should not be solved by apps that strive to do NAT traversals.
- full cone weakens firewall security.
- for apps that do not do NAT traversals, more general options exist, like VPNs (tinc, tailscale, zerotier, etc).

PS.
i think i've read somewhere that linux doesn't support full cone. (i guess that's a testament to its usefulness.) userland doesn't worry me much. but introducing a kernel patch to an already complex kernel function -and having to maintain it- to support a feature of questionable value is IMHO ill advised. if/when full cone is attainable with userland-only code, i'd still not work towards it, but would no longer oppose it.

**Comment by lscorcia** [comment]
Could you consider picking PR #10238 (already merged to master) into 22.03? The performance benefits are quite noticeable on mt7621.

**Comment by jwhb** [comment]
[NAMIB project](https://github.com/namib-project/) is looking forward for inclusion of dnsmasq v2.87test9+ for recently added MUD/RFC8520 support.

**Comment by jospezial** [comment]
https://openwrt.org/releases/22.03/notes-22.03.0

https://github.com/openwrt/openwrt/releases/tag/v22.03.0
I don't see it in https://github.com/openwrt/openwrt/releases
Is a release flag missing?

**Comment by mpratt14** [comment]
please list #9507 as a "known issue"

**Comment by lespinasse** [comment]
Please also list https://github.com/openwrt/openwrt/issues/10243 as known issue, remaining in 22.03.0 final. MediaTek MT7610E 5GHz radio is not functional, at least on tl-wr902ac, but I have a suspicion that all mt76 radios other than mt7615, mt7915 and mt7921 might be broken.

**Comment by Sami32** [comment]
What about adding #6329 ?

**Comment by escape0707** [comment]
> > snat + dnat = port forward != fullcone nat
> 
> That still does not explain what "fullcone NAT" _is_. From what I could gather in the net, it is basically a form of NAT where a specific internal port of a specific internal host is mapped to a unique external port. A traditional port forward ensures that a given external port is forwarded to a specific internal host and port while egress traffic uses a random external port auto-allocated by _MASQUERADE_. Adding an additional SNAT rule to explicitly rewrite _outbound_ traffic from the internal host and port to the specific external port without relying on MASQUERADE egress port auto allocation will achieve exactly that. What else is needed to achieve "full cone NAT" ?
> 
> In other words it could be as simple as emitting an additional `-s internalip --sport internalport -j SNAT --to-source externalip:externalport` ipt rule in addition to the `--dport externalport -j DNAT --to-destination internalip:internalport` one. This could likely be even done automatically by the firewall (maybe hidden behind a config knob).
> 
> Is that what you mean with "full cone NAT"?

Since my issue about miniupnpd is included within this release requirement, and less of a full cone implementation is the reason why I have to use UPnP in the first place, please allow me to add my words.

I mean, I feel bad for that user who got blocked from OpenWrt project. He is a true user who don't necessarily possess all the skills and knowledge to grasp exactly what the specifications are and which code pieces are needed to fix for his requirement, but he actively tried to convey the idea and did his research.

Even [the famous Tor project use the term "full-cone NAT"](https://community.torproject.org/relay/setup/snowflake/standalone/), and the RFC-4787 clearly described [the definition and benefit for having an endpoint-independent-{mapping, filtering}](https://www.rfc-editor.org/rfc/rfc4787#section-5). People have been asking why OpenWrt won't support full cone NAT since you, **jow-**, denied them [almost four years ago](https://github.com/openwrt/openwrt/issues/7021#issuecomment-1040297034) while still asking them for "what is the full cone NAT you are talking about?" 

> PS.
> i think i've read somewhere that linux doesn't support full cone. (i guess that's a testament to its usefulness.) userland doesn't worry me much. but introducing a kernel patch to an already complex kernel function -and having to maintain it- to support a feature of questionable value is IMHO ill advised. if/when full cone is attainable with userland-only code, i'd still not work towards it, but would no longer oppose it.

Most carrier home router and **all CGNAT DO** support that concept out of the box, so don't say that simply because Linux doesn't support it in its PC / server OS kernel means it's some shitty concept for you guys to even bother to learn about after users mentioned it through all these years.

In conclusion, the so-called "full cone NAT" = "Endpoint-Independent Mapping" + "Endpoint-Independent Filtering", as stated in the RFC 4787. And "DNAT + SNAT" doesn't bring this behavior out of the box / automated. Sometimes users don't even have the knowledge about which ports their program would use for hole punching. I don't mean full cone must be implemented in the near future or at all. But the current state of the resulting consensus simply make me feel grieving and depressed.

See also https://forum.openwrt.org/t/openwrt-xbox-call-of-duty-and-open-nat/67348

**Comment by fanthos** [comment]
Fullcone should be blocker for releases, but it is a good feature for many users.
Cone nat mostly used with udp protocol, replaces `(lanip, lanport, natport, wanip, wanport)` to `(lanip, lanport, natport)`, so same `lanip:lanport` will get (almost) always same port on gateway. It is very useful for p2p downloading or p2p gaming. Unlike upnp, upnp can map other lan ip or port so there is security risk, but cone nat only works for the port already opened.

**Comment by Shine-** [comment]
Beside the point that I think this is completely off-topic here.

As far as I understood the term "full cone NAT", it simply means - as others already correctly stated above - that `internal-src-ip:internal-src-port` will always be NATed to the same (either predefined or reserved-on-first-use) `puplic-src-ip:public-src-port`, instead of dynamically and based on availablility like any normal NAT. From my experience, this is only needed for pain-in-the-ass protocols like VoIP and other realtime bs.

Coining a term as "full", trying to create the impression that everyting else is "not full" (ie. incomplete), doesn't actually make this true. "Full cone NAT" is, in fact, the severely limited version of NAT, as it requires reserving precious source ports close-to-forever.

IMHO, solving design problems for shitty protocols should not be the endpoint router's responsibility. The **less** endpoint routers support this bs, the faster it will be solved on protocol resp. implementation side.

In fact, e.g. most SIP servers/proxies do already solve this protocol design flaw in their implementation by ignoring any IP and port provided in the endpoint's SIP registration message, and using the actual source IP and port they receive the connection from, while requiring frequent keepalive messages, in order to to get updated with any changed IP or port asap. More modern implementations don't even use archaic protocols with major design flaws at all. Which is the way forward.

But in case you **really** need "full cone NAT" (damn I hate this term), just see the above comments for examples on how to add it manually using ipt or nft. It's not like OpenWRT doesn't support it, you know?

**Comment by escape0707** [comment]
> As far as I understood the term "full cone NAT", it simply means - as others already correctly stated above - that `internal-src-ip:internal-src-port` will always be NATed to the same (either predefined or reserved-on-first-use) `puplic-src-ip:public-src-port`, instead of dynamically and based on availablility like any normal NAT. From my experience, this is only needed for pain-in-the-ass protocols like VoIP and other realtime bs.

You are wrong. What you are saying is exactly like port-forwarding or {S,D}NAT joe- proposed as a "solution". Endpoint-independent-filtering is not that. Endpoint-independent-mapping is kinda close but is still not what you thought that internal ip:port will always be NATed to the same public ip:port. Please give a look about the RFC I've mentioned and don't make the conversation more noisy. I'm not willing to stress it again and again, but people denying this will simply not read [4.1.  Address and Port Mapping](https://www.rfc-editor.org/rfc/rfc4787#section-4.1) and [5.  Filtering Behavior](https://www.rfc-editor.org/rfc/rfc4787#section-5).

**Comment by Shine-** [comment]
From the RFC:
> Endpoint-Independent Mapping (EIM):  A term applied to NAT operation
      where all mappings from a single internal address, protocol, and
      port to different remote peers and ports are all assigned the same
      external address and port.

Which is exactly what I wrote above, in different words. I end this here. Off-topic remains off-topic.

**Comment by fanthos** [comment]
The different between cone nat and d/snat command is, ports assign in Cone nat is dynamic, not like manually added static port mapping with SNAT and DNAT cmdline.

**Comment by escape0707** [comment]
Okay, let me change my first comment. I think I stated a wrong statement there, that "fullcone nat" is not about Endpoint-independent-mapping as during the UDP hole punching, the internal host will not simple "close" the first session and "start" another, it's stateless.

For better gaming and p2p hole punching, only the "endpoint-independent-filtering" is needed. And that should be all those users requesting "fullcone nat" again and again wanted. The mapping part is usually not a problem, ig.

**Comment by Shine-** [comment]
Off-topic as in: You want a global change in nftables behavior, as opposed to having to define it in a rule set where needed. Why do you request this from OpenWRT, and not from nftables in the first place?

**Comment by escape0707** [comment]
> I don't mean full cone must be implemented in the near future or at all. But the current state of the resulting consensus simply makes me feel grieving and depressed.

I'm not requesting anything. I just want to clarify what the "ghost" user, fanthos and people [in this thread](https://forum.openwrt.org/t/openwrt-xbox-call-of-duty-and-open-nat/67348) want.

And nftable might not need this feature so badly, but a lot of OpenWrt users do. They even came up of work around before OpenWrt migrated to nftables. I think ghost already linked many of them here.

**Comment by Shine-** [comment]
Create a workaround, PR it, and see what the devs think about it. It wouldn't be the first time that OpenWRT implemented workarounds for stuff that upstream didn't accept.

But - mind you, that's purely my personal opinion - this pinned GH issue was meant for addressing blocking issues that urgently need a fix/implementation before release, and for bugs that already have a fix, just not merged yet. Not for random feature requests that have never been on the agenda for 22.03 before.

**Comment by escape0707** [comment]
I don't want to add more comment. I think my first comment after appied the edition you pointed as wrong, is sufficient to clear up any previous misunderstanding. I don't want to discuss about this problem throughly here either.

But do remember that this is not "random feature requests that have never been on the agenda for 22.03 before". As if you clearly read what I clarified as "endpoint-independent-filtering"'s benifit and why people want it and how they got denied for ["requesting"](https://github.com/openwrt/openwrt/issues/7021) and "PR" them, the conversation won't need to extend beyond my first comment.

**Comment by skbeh** [comment]
> Why do you request this from OpenWRT, and not from nftables in the first place?

Because you can not define a rule set in real time without [integrating into fw4](https://github.com/wongsyrone/lede-1/blob/master/package/network/config/firewall4/patches/999-01-firewall4-add-fullcone-support.patch) and netfilter, which requires at least [a openwrt package](https://github.com/wongsyrone/lede-1/blob/master/package/external/nft-fullcone/Makefile) to build a kernel module with the kernel.

**Comment by escape0707** [comment]
I just finished reading RFC 5128 which concluded about hole punching and is using a quite different way than I imagined. That only need "endpoint-independent-mapping" on both peers' NAT to achieve UDP hole punching elegantly. But sadly many games and applications nowadays didn't implement it right as still need "endpoint-independent-filtering". I also read about some parts of BCP 127, RFC 2663, 3489, 4787, 5382, again and again because you don't want to. **I've edited all my comments to reflect the new knowledge and refreshment I got, and make things easier to understand. Please revise again, if you care.**

Thank you guys and god, I'm officially not any kind of NAT enjoyer but more like an RFC / referencing enjoyer now! (Just kidding, I feel bullied).

**Comment by jow-** [comment]
Please take the full cone NAT discussion upstream to netfilter and try to get it included in official kernel and nftables releases, then it will appear in OpenWrt automatically. If the feature is really widely required and important, it should get into netfilter proper and its implementation scrutinized by the kernel/netfilter developers. As it stands, OpenWrt lacks ressources and motivation to maintain this out-of-tree feature.

**Comment by fanthos** [comment]
After some research, the nf-nat flag `persistent` means for mulitple ip setup, use same src ip for same dest ip.
I found out of tree impl of cone nat at https://github.com/fullcone-nat-nftables/
